File: resources/[np-framework]/np-lib/client/cl_ui.lua
local registered = {}

function RegisterUICallback(name, cb)
  local function interceptCb(data, innerCb)
    cb(data, function(result)
      if result.meta.ok then
        result.meta.message = "done"
      end
      innerCb(result)
    end)
  end
  AddEventHandler(('_npx_uiReq:%s'):format(name), interceptCb)

  if (GetResourceState("np-ui") == "started") then
    pcall(function() exports["np-ui"]:RegisterUIEvent(name) end)
  end

  registered[#registered + 1] = name
end

function SendUIMessage(data)
  if (GetResourceState("np-ui") == "started") then
    pcall(function() exports["np-ui"]:SendUIMessage(data) end)
  end
end

function SetUIFocus(hasFocus, hasCursor)
  if (GetResourceState("np-ui") == "started") then
    pcall(function() exports["np-ui"]:SetUIFocus(hasFocus, hasCursor) end)
  end
end

function GetUIFocus()
  if (GetResourceState("np-ui") == "started") then
    local ok, res = pcall(function() return exports["np-ui"]:GetUIFocus() end)
    if ok then return res end
  end
  return false
end

AddEventHandler("_npx_uiReady", function()
  for _, eventName in ipairs(registered) do
    if (GetResourceState("np-ui") == "started") then
      pcall(function() exports["np-ui"]:RegisterUIEvent(eventName) end)
    end
  end
end)

File: resources/[np-framework]/np-lib/client/cl_ui.js
// np-lib: cl_ui.js
const registered = [];

function RegisterUICallback(name, cb) {
    AddEventHandler(`_npx_uiReq:${name}`, cb);

    try {
        if (GetResourceState('np-ui') === 'started' && exports['np-ui'] && exports['np-ui'].RegisterUIEvent) {
            exports['np-ui'].RegisterUIEvent(name);
        }
    } catch (e) {}

    registered.push(name);
}

function SendUIMessage(data) {
    try {
        if (exports['np-ui'] && exports['np-ui'].SendUIMessage) {
            exports['np-ui'].SendUIMessage(data);
        }
    } catch (e) {}
}

function SetUIFocus(hasFocus, hasCursor) {
    try {
        if (exports['np-ui'] && exports['np-ui'].SetUIFocus) {
            exports['np-ui'].SetUIFocus(hasFocus, hasCursor);
        }
    } catch (e) {}
}

function GetUIFocus() {
    try {
        if (exports['np-ui'] && exports['np-ui'].GetUIFocus) {
            return exports['np-ui'].GetUIFocus();
        }
    } catch (e) {}
    return false;
}

AddEventHandler('_npx_uiReady', () => {
    try {
        if (exports['np-ui'] && exports['np-ui'].RegisterUIEvent) {
            registered.forEach((eventName) => exports['np-ui'].RegisterUIEvent(eventName));
        }
    } catch (e) {}
});
